internal/worker/concurrency_auto_scaler.go

```diff
@@ -0,0 +1,263 @@
+// Copyright (c) 2017-2021 Uber Technologies Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package worker
+
+import (
+	"math"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/jonboulle/clockwork"
+	"github.com/uber-go/tally"
+	"go.uber.org/zap"
+
+	"go.uber.org/cadence/.gen/go/shared"
+)
+
+const (
+	defaultAutoScalerUpdateTick = time.Second
+	// concurrencyAutoScalerObservabilityTick = time.Millisecond * 500
+	targetPollerWaitTimeInMsLog2  = 4 // 16 ms
```

Comment:
took a while to figure out what exactly the goal here was, I think because the log2/exp2 steps are so separated...

might be easier to follow if you make this a normal ms value, and do a `current*log2(actual/target)` instead?  feels more like a smoothing operation at that point / a "react to the magnitude rather than the value", where currently it's kinda hidden as a `current*log2(actual)/log2(target)` that's being done in three separate locations.

re `TimeInMs` vs a `time.Duration`: seems fine to me.  it's a ton of value-casting noise otherwise, sadly.  and/or maybe we should consider a "duration math" util somewhere, as we do a moderate amount of duration-related floaty calculations in both client and server, and they're super verbose.